python/lib/modeldata/dmod/modeldata/data/object_store_manager.py

```diff
@@ -292,8 +302,22 @@ def add_data(self, dataset_name: str, dest: str, domain: DataDomain, data: Optio
             msg = "{}.{} source path '{}' does not exist."
             raise ValueError(msg.format(self.__class__.__name__, _getframe(0).f_code.co_name, source))
         elif src_path.is_dir():
-            bucket_root = kwargs.get('bucket_root', src_path)
-            self._push_files(bucket_name=dataset_name, dir_path=src_path, bucket_root=bucket_root)
+            # Recognized scenario when we may want to take advantage of small file archives
+            # (see https://blog.min.io/small-file-archives/)
+            # Also, we already know from above that, if read-only, dataset must also be empty
+            # TODO: (later) consider whether there is some minimum file count (perhaps configurable) to also consider
+            if self.datasets[dataset_name].is_read_only:
```

Comment:
This feels like we are injecting a use case specific feature into a general api. For instance, in creating a forcing dataset I would likely mark it as read-only as it is more or less static. I don't think, in that case, it is desirable to zip up a directory of, lets say, netcdf forcing files. Zipping up a directory seems like it should be handled by the specific application and not this api. We will have to rethink how the `data_archiving` attribute is set on a `Dataset` if that is the case though.

One thought is to create an algebraic datatype like a `DataFormat.ARCHIVE[ArchiveFormat, DataFormat]` that is, itself, a `DataFormat` but also wraps an `ArchiveFormat` and a `DataFormat`. It might be a little cumbersome to capture this using a python `Enum`, but we can certainly figure out how to best express the idea in the type system. This could be used to set if a `Dataset` is `data_archiving` or not if it still makes sense to keep that as a top level attribute.

This will move complexity to other places, but I think retain some desirable traits of the system more generally. Im sure there is a simpler way to capture this idea that what ive suggested. I think we should strive for simplicity until we can't.

Reply:
> This feels like we are injecting a use case specific feature into a general api ...

I disagree with the headline here but agree with parts of your argument in isolation.  

It is a use case specific feature, but it's not part of the API.  It's a specific implementation of the API that provides this behavior, but behind the scenes and particular to a subset of use cases.  I don't see a better way to introduce storage-backend-specific optimizations other than doing it within the thing interacting with the backend.  A specific application is not going to have any idea (nor should it) that there is a performance penalty for certain write scenarios (e.g., lots of files to the object store).

And, just to be clear, this is intended as a backend-specific optimization.  It's only being introduced _here_ because MinIO can take advantage of ZIP files in a particular way.  I don't have a problem with other things _also_ archiving data in a dataset in the future, if they have other reasons to do so.  I expect eventually we'll want to; hence, the attribute within _Dataset_ instead of just tracking inside the manager.   But those are apples and oranges.

But I agree that this doesn't consider all scenarios sufficiently:  a single netcdf forcing file, for example, probably doesn't need to be archive.  Probably, 50 forcing CSV files don't really either.  I even put in a TODO comment to this effect, but didn't want to just throw out a magic number of files for the too-many threshold.  I'm open to discussing either what that number should be or how to better determine when the object store manager really can/should do this.

> One thought is to create an algebraic datatype like a DataFormat.ARCHIVE[ArchiveFormat, DataFormat] that is, itself, a DataFormat but also wraps an ArchiveFormat and a DataFormat

That seems much more complicated and far reaching that what's in place now.  And I still don't think we could effectively apply it without making the things creating the data start to worry too much about the details of the data storage backend\*.  Not to mention it would convolute logic for reconciling data requirements.  

\* Caveat that this is already the case for workers writing output, kind of, though they don't interact with the DMOD data orchestration in the same way, so their data writing is written more in isolation (at least for now).